### PR TITLE
ClusterDeployment and CertificateRequest Reconcile Prometheus Metrics

### PR DESCRIPTION
**Pull request by dustman9000**
This pull request adds two new histogram custom metrics for observability into reconcile performance of ClusterDeployment and CertificateRequest CR reconciliation.

Tested locally with a dummy ClusterDeployment and subsequent CertificateRequest:
```
$ kubectl run -it --rm debug --image=radial/busyboxplus:curl --restart=Never -- curl http://172.17.0.3:8080/metrics
# HELP certman_operator_certificate_issue_duration Runtime of issue certificate function in seconds
# TYPE certman_operator_certificate_issue_duration summary
certman_operator_certificate_issue_duration_sum{name="certman-operator"} 0
certman_operator_certificate_issue_duration_count{name="certman-operator"} 0
# HELP certman_operator_certificate_request_reconcile_duration_seconds The duration it takes to reconcile a CertificateRequest
# TYPE certman_operator_certificate_request_reconcile_duration_seconds histogram
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="0.005"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="0.01"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="0.025"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="0.05"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="0.1"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="0.25"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="0.5"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="1"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="2.5"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="5"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="10"} 2
certman_operator_certificate_request_reconcile_duration_seconds_bucket{name="certman-operator",le="+Inf"} 2
certman_operator_certificate_request_reconcile_duration_seconds_sum{name="certman-operator"} 0.0036778180000000002
certman_operator_certificate_request_reconcile_duration_seconds_count{name="certman-operator"} 2
# HELP certman_operator_cluster_deployment_reconcile_duration_seconds The duration it takes to reconcile a ClusterDeployment
# TYPE certman_operator_cluster_deployment_reconcile_duration_seconds histogram
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="0.005"} 0
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="0.01"} 0
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="0.025"} 0
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="0.05"} 0
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="0.1"} 0
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="0.25"} 1
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="0.5"} 1
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="1"} 1
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="2.5"} 1
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="5"} 1
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="10"} 1
certman_operator_cluster_deployment_reconcile_duration_seconds_bucket{name="certman-operator",le="+Inf"} 1
certman_operator_cluster_deployment_reconcile_duration_seconds_sum{name="certman-operator"} 0.103815477
certman_operator_cluster_deployment_reconcile_duration_seconds_count{name="certman-operator"} 1
```

### CODE REVIEW

**Comment by dustman9000** [comment]
> The way we've done this in the other operators was to use a `HistogramVec` with a `label` for the controller. Is there a reason we need two separate metrics?

It's easier to for the controllers to explicitly call the functions associated with their metrics. Using a label and lookup is unnecessary complexity.

**Comment by 2uasimojo** [comment]
> > The way we've done this in the other operators was to use a `HistogramVec` with a `label` for the controller. Is there a reason we need two separate metrics?
> 
> It's easier for the controllers to explicitly call the functions associated with their metrics. Using a label and lookup seems like unnecessary added complexity. This is also more consistent with how the localmetrics package is being used in this operator.

Would you suggest we refactor the other projects to do the same? It would be nice to be standardized to some extent. Including how consumers expect to retrieve the metrics (e.g. for graphing purposes).

**Comment by dustman9000** [comment]
> > > The way we've done this in the other operators was to use a `HistogramVec` with a `label` for the controller. Is there a reason we need two separate metrics?
> > 
> > 
> > It's easier for the controllers to explicitly call the functions associated with their metrics. Using a label and lookup seems like unnecessary added complexity. This is also more consistent with how the localmetrics package is being used in this operator.
> 
> Would you suggest we refactor the other projects to do the same? It would be nice to be standardized to some extent. Including how consumers expect to retrieve the metrics (e.g. for graphing purposes).

I agree standardization is important, but this operator approaching EOL soon. This operator requires much more refactoring and rework to bring it up to speed beyond just Histogram vs. HistrogramVec metrics. As I said, it's a bit unnecessary and outside the scope of this PR.

File: cmd/manager/main.go [hunk 1]
```diff
@@ -97,6 +97,8 @@ func main() {
 	// Set default manager options
 	options := manager.Options{
 		Namespace: namespace,
+		// Disable controller-runtime metrics serving
+		MetricsBindAddress: "0",
```
Comment (yithian) [inline cmd/manager/main.go:101]: It didn't look like it from the output you pasted, but we won't be losing anything by disabling this?

Comment (dustman9000) [reply cmd/manager/main.go:101]: This is actually fixing a bug where our current custom metrics are being overridden by the controller-runtime metrics server. If you look at certman-operator on hive-stage there are no custom metrics being served. So in fact we are gaining, not disabling things here. 👍

Comment (yithian) [reply cmd/manager/main.go:101]: :+1:

**Pull request merged**